apollo-router/src/plugins/traffic_shaping/deduplication.rs

```diff
@@ -57,23 +68,47 @@ impl Clone for CloneSubgraphResponse {
 pub(crate) struct QueryDeduplicationService<S: Clone> {
     service: S,
     wait_map: WaitMap,
+    deduplicate_query_ignored_headers: Option<Vec<String>>,
 }
 
 impl<S> QueryDeduplicationService<S>
 where
     S: tower::Service<SubgraphRequest, Response = SubgraphResponse, Error = BoxError> + Clone,
 {
-    fn new(service: S) -> Self {
+    fn new(service: S, deduplicate_query_ignored_headers: Option<Vec<String>>) -> Self {
         QueryDeduplicationService {
             service,
             wait_map: Arc::new(Mutex::new(HashMap::new())),
+            deduplicate_query_ignored_headers,
         }
     }
 
+    fn get_cache_key(
+        request: SubgraphRequest,
+        deduplicate_query_ignored_headers: Option<Vec<String>>,
+    ) -> (http_ext::Request<Request>, Arc<CacheKeyMetadata>) {
+        let authorization_cache_key = request.authorization.clone();
+        let cache_key: (http_ext::Request<Request>, Arc<CacheKeyMetadata>) =
+            if let Some(headers) = deduplicate_query_ignored_headers {
+                let mut cloned_request = request.clone();
+                for header in headers {
+                    cloned_request.subgraph_request.headers_mut().remove(header);
+                }
+                (
+                    (cloned_request.subgraph_request).into(),
+                    authorization_cache_key,
+                )
+            } else {
+                ((&request.subgraph_request).into(), authorization_cache_key)
+            };
```

Comment:
I think to make something more performant we will have to change the cache_key itself and hash the whole request. A little bit like we did for subscription. And talking about subscription that would be great if we could replicate the same logics for subscription too. The main issue I can see here is we're cloning the http request and also ignored headers.

Reply:
I managed to get rid most of the cloning by passing borrowed values.

I'm not sure if I can avoid the borrow on line 93 here because I need to create a cache key that doesn't include that header without mutating the original request.

Open to suggestions though!

Reply:
I agree with @bnjjj here. We don't need to keep the entire request as key, it can be a hash. That will be shorter, fastrr to compare, use less memory etc

Reply:
Okay, awesome. I rewrote this to use a hash of the request + a hash of the authorization context!